From 274c8e532c6fc30bb680661f420984e1031814cf Mon Sep 17 00:00:00 2001
From: Randolph <randolph@andestech.com>
Date: Mon, 3 Mar 2025 18:37:03 +0800
Subject: [PATCH 53/53] swdtb: support APLIC as the interrupt controller [Bug
 33626] (#164)

In the previous implementation of l2c_interrupt_fixup(), the default
interrupt controller was PLIC. After invoking the AIA, the interrupt
controller changes to APLIC.

Signed-off-by: Randolph <randolph@andestech.com>

Reviewed-on: https://gitea.andestech.com/RD-SW/u-boot.official/pulls/164
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: Alex Chun-Ju Lin <alex749@andestech.com>
Reviewed-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Co-authored-by: Randolph <randolph@andestech.com>
Co-committed-by: Randolph <randolph@andestech.com>
---
 .../common/swdtb_merger/swdtb_merger_fixup.c  | 17 +++++-------
 .../common/swdtb_merger/swdtb_tool.c          | 26 +++++++++++++++++++
 .../common/swdtb_merger/swdtb_tool.h          |  1 +
 3 files changed, 33 insertions(+), 11 deletions(-)

diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
index 7809a2d9a0d4..9037f50e3298 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
@@ -407,7 +407,6 @@ int l2c_interrupt_fixup(void *fdt, int node, char *name, char *value)
 	u32 target_pin = 16;
 	u32 target_trig = IRQ_TYPE_LEVEL_HIGH;
 
-	const char *irq_compat = "riscv,plic0";
 	const char *irq_cell_str = "#interrupt-cells";
 	int irq_node;
 	int irq_phandle;
@@ -428,13 +427,9 @@ int l2c_interrupt_fixup(void *fdt, int node, char *name, char *value)
 
 		show_rawdtb(target, target_node);
 
-		/* Add interrupt-parent */
-		irq_node = fdt_node_offset_by_compatible(fdt, 0, irq_compat);
-		if (irq_node < 0) {
-			dtb_error("Error on fdt_node_offset_by_compatible: %s\n",
-				  fdt_strerror(irq_node));
-			return irq_node;
-		}
+		irq_node = find_interrupt_node(fdt);
+		if (!irq_node)
+			return !(RETURN_DONE);
 
 		irq_phandle = fdt_get_phandle(fdt, irq_node);
 		if (!irq_phandle) {
@@ -442,7 +437,7 @@ int l2c_interrupt_fixup(void *fdt, int node, char *name, char *value)
 				  fdt_strerror(irq_phandle));
 		}
 
-		/* Add interrupts */
+		/* For checking cell is 2 */
 		irq_prop = fdt_getprop(fdt, irq_node,
 				       irq_cell_str, NULL);
 
@@ -459,7 +454,7 @@ int l2c_interrupt_fixup(void *fdt, int node, char *name, char *value)
 			return -EINVAL;
 		}
 
-		/* For cell is 2 */
+		/* Add interrupt-parent */
 		ret = fdt_appendprop_u32(fdt, node,
 					 target_parent, irq_phandle);
 		if (ret < 0) {
@@ -470,7 +465,7 @@ int l2c_interrupt_fixup(void *fdt, int node, char *name, char *value)
 
 		tmp[0] = cpu_to_fdt32(target_pin);
 		tmp[1] = cpu_to_fdt32(target_trig);
-
+		/* Add interrupts */
 		ret = fdt_appendprop(fdt, node, target,
 				     tmp, sizeof(tmp));
 		if (ret) {
diff --git a/board/andestech/common/swdtb_merger/swdtb_tool.c b/board/andestech/common/swdtb_merger/swdtb_tool.c
index 374321a257bf..155b777029e7 100644
--- a/board/andestech/common/swdtb_merger/swdtb_tool.c
+++ b/board/andestech/common/swdtb_merger/swdtb_tool.c
@@ -411,3 +411,29 @@ int fdt_node_name_cmp(const void *fdt, int node, const char *target_node)
 	else
 		return 0;
 }
+
+int find_interrupt_node(const void *fdt)
+{
+	const char *irq_compat = "riscv,plic0";
+	const char *aplic_compat = "riscv,aplic";
+	const char *m_children = "riscv,children";
+	int irq_node;
+	const fdt32_t *s_phandle;
+
+	/* Find the interrupt controller node using plic */
+	irq_node = fdt_node_offset_by_compatible(fdt, 0, irq_compat);
+	if (irq_node > 0)
+		return irq_node;
+
+	/* Find the interrupt controller node using aplic */
+	fdt_for_each_node_by_compatible(irq_node, fdt, -1, aplic_compat) {
+		/* If the node has "riscv,children", it's m-mode aplic */
+		s_phandle = fdt_getprop(fdt, irq_node, m_children, NULL);
+		if (s_phandle)
+			continue;
+		return irq_node;
+	}
+
+	dtb_error("Cannot find the interrupt\n");
+	return 0;
+}
diff --git a/board/andestech/common/swdtb_merger/swdtb_tool.h b/board/andestech/common/swdtb_merger/swdtb_tool.h
index aae107c2f7be..e2f2fbb4b6fa 100644
--- a/board/andestech/common/swdtb_merger/swdtb_tool.h
+++ b/board/andestech/common/swdtb_merger/swdtb_tool.h
@@ -106,4 +106,5 @@ char *replace_serial_string(const char *origin_string, int replace_start,
 			    int new_length);
 
 int fdt_node_name_cmp(const void *fdt, int node, const char *target_node);
+int find_interrupt_node(const void *fdt);
 #endif
-- 
2.34.1

